src/transformStateWithClones.js

```diff
@@ -7,7 +7,50 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const ADD_PROPERTY_TYPE = 'addProperties';
+  const REMOVE_PROPERTY_TYPE = 'removeProperties';
+  const CLEAR_TYPE = 'clear';
+
+  const stateCopy = { ...state };
+  const stateArray = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case ADD_PROPERTY_TYPE:
+        for (const key of Object.keys(action.extraData)) {
+          stateCopy[key] = action.extraData[key];
+        }
+
+        break;
+
+      case REMOVE_PROPERTY_TYPE:
+        for (let j = 0; j < action.keysToRemove.length; j++) {
+          const key = action.keysToRemove[j];
+
+          if (stateCopy.hasOwnProperty(key)) {
+            delete stateCopy[key];
+          }
+        }
+
+        break;
+
+      case CLEAR_TYPE:
+        for (const key in stateCopy) {
+          if (stateCopy.hasOwnProperty(key)) {
```

Comment:
This check doesn't make sense because you're iterating over the object's properties and then checking to see if it contains them